File: builtin/net.js
var tm = process.binding('tm');

var util = require('util');
var Stream = require('stream').Stream;

/**
 * ssl
 */

var ssl_ctx = null;

function ensureSSLCtx () {
  if (ssl_ctx == null) {
    ssl_ctx = tm.ssl_context_create();
  }
}


/**
 * TCPSocket
 */

function TCPSocket (socket, _secure) {
  this.socket = socket;
  this._secure = _secure;
}

util.inherits(TCPSocket, Stream);

TCPSocket.prototype.connect = function (port, ip, cb) {
  var ips = ip.split('.');
  var self = this;
  setImmediate(function () {
    tm.tcp_connect(self.socket, Number(ips[0]), Number(ips[1]), Number(ips[2]), Number(ips[3]), Number(port));
    
    if (self._secure) {
      var ssl = tm.ssl_session_create(ssl_ctx, self.socket);
      self._ssl = ssl;
    }

    self.__listen();
    cb();
    self.emit('connect');
  });
};

TCPSocket.prototype.__listen = function () {
  var self = this;
  this.__listenid = setInterval(function () {
    var buf = '';
    while (self.socket != null && tm.tcp_readable(self.socket) > 0) {
      if (self._ssl) {
        var data = tm.ssl_read(self._ssl);
      } else {
        var data = tm.tcp_read(self.socket);
      }
      if (!data || data.length == 0) {
        break;
      }
      buf += data;
    }
    if (buf.length) {
      self.emit('data', buf);
    }
  }, 0);
};

TCPSocket.prototype.write = function (buf, cb) {
  var self = this;
  setImmediate(function () {
    if (self._ssl) {
      tm.ssl_write(self._ssl, buf, buf.length);
    } else {
      tm.tcp_write(self.socket, buf, buf.length);
    }
    if (cb) {
      cb();
    }
  })
};

TCPSocket.prototype.destroy = TCPSocket.prototype.close = function () {
  var self = this;
  if (this.__listenid != null) {
    clearInterval(this.__listenid);
    this.__listenid = null
  }
  setImmediate(function () {
    tm.tcp_close(self.socket);
    self.socket = null;
    self.emit('close');
  });
};

exports.connect = function (port, host, callback, _secure) {
  if (_secure) {
    ensureSSLCtx();
  }

  var sock = tm.tcp_open();
  if (sock == -1) {
    throw 'ENOENT: Cannot connect to new socket.'
  }
  
  var client = new TCPSocket(sock, _secure);
  client.connect(port, host, callback);
  return client;
};


/**
 * Server
 */

function TCPServer (socket) {
  TCPSocket.call(this, socket);
}

util.inherits(TCPServer, TCPSocket);

TCPServer.prototype.listen = function (port, ip) {
  var self = this;
  var res = tm.tcp_listen(this.socket, port);
  if (res < 0) {
    throw "Error listening on TCP socket (port " + port + ", ip " + ip + ")"
  }

  setInterval(function () {
    var client;
    if ((client = tm.tcp_accept(self.socket)) >= 0) {
      var clientsocket = new TCPSocket(client);
      clientsocket.__listen();
      self.emit('socket', clientsocket);
    }
  });
};

exports.createServer = function (onsocket) {
  var server = new TCPServer(tm.tcp_open());
  onsocket && server.on('socket', onsocket);
  return server; 
};

File: builtin/net.lua
function (_ENV)
local string, math, print, type, pairs = nil, nil, nil, nil, nil;
local _module = _obj({exports=_obj({})}); local exports, module = _module.exports, _module;

local util, Stream, TCPSocket, TCPServer = util, Stream, TCPSocket, TCPServer;
TCPSocket = (function (this, socket)
local TCPSocket = _debug.getinfo(1, 'f').func;
(this).socket = socket;
end);
TCPServer = (function (this, socket)
local TCPServer = _debug.getinfo(1, 'f').func;
if TCPSocket:call(this, socket) then end;
end);
util = require(global, ("util"));
Stream = (require(global, ("stream"))).Stream;
if util:inherits(TCPSocket, Stream) then end;
((TCPSocket).prototype).connect = (function (this, port, ip, cb)
local ips, client = ips, client;
ips = ip:split(("."));
client = this;
if setImmediate(global, (function (this)
if tm__tcp__connect(global, client.socket, Number(global, ips[(0)]), Number(global, ips[(1)]), Number(global, ips[(2)]), Number(global, ips[(3)]), Number(global, port)) then end;
if (function () local base, prop = client, "__listen"; return base[prop](base); end)() then end;
if cb(global) then end;
end)) then end;
end);
((TCPSocket).prototype).__listen = (function (this)
local client = client;
client = this;
if setInterval(global, (function (this)
local buf = buf;
while _truthy((client).socket and (tm__tcp__readable(global, client.socket) > (0))) do

buf = tm__tcp__read(global, client.socket);
if _truthy((not (buf)) or ((buf).length == (0))) then
_c = _break; break;
end
if client:emit(("data"), buf) then end;

end
end), (100)) then end;
end);
((TCPSocket).prototype).write = (function (this, buf, cb)
local socket = socket;
socket = (this).socket;
if setImmediate(global, (function (this)
if tm__tcp__write(global, socket, buf, buf.length) then end;
if _truthy(cb) then
if cb(global) then end;
end
end))
 then end;
end);
((TCPSocket).prototype).close = (function (this)
local self = self;
self = this;
if setImmediate(global, (function (this)
if tm__tcp__close(global, self.socket) then end;
(self).socket = (null);
if self:emit(("close")) then end;
end)) then end;
end);
(exports).connect = (function (this, port, host, callback)
local client = client;
client = _new(TCPSocket, tm__tcp__open(global));
if client:connect(port, host, callback) then end;
if true then return client; end;
end);
if util:inherits(TCPServer, TCPSocket) then end;
((TCPServer).prototype).listen = (function (this, port, ip)
local self, res = self, res;
self = this;
res = tm__tcp__listen(global, this.socket, port);
if (res < (0)) then
_error(((((("Error listening on TCP socket (port ") + port) + (", ip ")) + ip) + (")")))
end
if setInterval(global, (function (this)
local client, clientsocket = client, clientsocket;
client = nil;
if _truthy((tm__tcp__readable(global, self.socket) > (0)) and (((function () local _r = tm__tcp__accept(global, self.socket); client = _r; return _r; end)()) >= (0))) then
clientsocket = _new(TCPSocket, client);
if (function () local base, prop = clientsocket, "__listen"; return base[prop](base); end)() then end;
if self:emit(("socket"), clientsocket) then end;
end
end)) then end;
end);
(exports).createServer = (function (this, onsocket)
local server = server;
server = _new(TCPServer, tm__tcp__open(global));
if onsocket and server:on(("socket"), onsocket) then end;
if true then return server; end;
end);

return _module.exports;
end